cognify/optimizer/core/unified_layer_opt.py

```diff
@@ -560,83 +539,90 @@ def _optimize_iteration(
-                                    and counter % self.save_ckpt_interval == 0
-                                ):
-                                    self.save_ckpt(
-                                        opt_config.opt_log_path,
-                                        opt_config.param_save_path,
-                                    )
-                                _update_pbar(pbar)
-                            if _should_exit() or self._should_stop:
-                                executor.shutdown(wait=False, cancel_futures=True)
-                                break
-                        except Exception as e:
-                            logger.error(f"Error in evaluating task: {e}", exc_info=True)
-                            raise
-
-        release_position(pbar_position)
+
+        if self.hierarchy_level == 0:
+            pbar.init_pbar(
+                total=float(num_current_trials + opt_config.n_trials),
+                initial=float(num_current_trials),
+                initial_score=initial_score,
+                initial_cost=initial_cost,
+                initial_exec_time=initial_exec_time,
+                opt_cost=self.opt_cost
+            )
+
+        counter = 0
+        if opt_config.throughput == 1:
+            for _ in range(opt_config.n_trials):
+                if _should_exit() or self._should_stop:
+                    break
+                result = self._optimize_iteration(base_program, frac / opt_config.n_trials)
+
+                if result is None or not result.complete:
+                    continue
+                counter += 1
+                if (
+                    self.save_ckpt_interval > 0
+                    and counter % self.save_ckpt_interval == 0
+                ):
+                    self.save_ckpt(
+                        opt_config.opt_log_path, opt_config.param_save_path
+                    )
+                _update_pbar(frac / opt_config.n_trials)
+        else:
+            with ThreadPoolExecutor(max_workers=opt_config.throughput) as executor:
+                futures = [
+                    executor.submit(self._optimize_iteration, base_program, frac / opt_config.n_trials)
+                    for _ in range(opt_config.n_trials)
+                ]
+                for f in as_completed(futures):
+                    try:
+                        result = f.result()
+                        if result and result.complete:
+                            counter += 1
+                            if (
+                                self.save_ckpt_interval > 0
+                                and counter % self.save_ckpt_interval == 0
+                            ):
+                                self.save_ckpt(
+                                    opt_config.opt_log_path,
+                                    opt_config.param_save_path,
+                                )
+                            _update_pbar(frac / opt_config.n_trials)
```

Comment:
how does frac here aggregate progress from multiple parallel execution?

Reply:
Actually, this function is misleading. It doesn't need `frac`. I've replaced it with the real call, which is to update the status bar description ("best xxx, total xxx"). 

Basically, `frac` is used at the individual data point level to move the progress bar. We handle parallel execution by acquiring a lock on the pbar when the request finishes. This part does not really add overhead. 

Then, at the end of each trial, we update the status bar description. I've tried updating it as frequently as the progress, but this adds lots of overhead.